plugins/inputs/ctrlx_datalayer_sse/ctrlx_datalayer_sse.go

```diff
@@ -0,0 +1,410 @@
+	}
+	err = json.Unmarshal(body, &vt)
+	if err != nil {
+		return false, fmt.Errorf("failed to unmarshal validate token response: %v", err)
+	}
+	return vt.Valid, nil
+}
+
+// setAuthToken gets a valid authentication token from the ctrlX CORE using username and password.
+// The token is necessary to create a subscription. An error will be returned, when no token could get aquired.
+func (s *CtrlXDataLayerSSE) setAuthToken() error {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+
+	v, err := s.checkToken()
+	if err != nil {
+		// Something went wrong during token validation
+		s.Log.Error(err)
+	}
+	if v {
+		// Token was already created and token still valid
+		return nil
+	}
+
+	tokenURL := combineScheme(s.Hostname) + identityPath
+	payload, err := json.Marshal(map[string]string{
+		"name":     s.Username,
+		"password": s.Password,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create authorization token request: %v", err)
+	}
+	requestBody := bytes.NewBuffer(payload)
+
+	resp, err := s.connection.Post(tokenURL, "application/json", requestBody)
+	if err != nil {
+		r := strings.NewReplacer(identityPath, "/...")
+		b := r.Replace(err.Error())
+		return fmt.Errorf("an error occured during request of authorization token: %v", b)
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("error while reading the response bytes: %v", err)
+	}
+
+	err = json.Unmarshal(body, &s.token)
+	if err != nil {
+		return fmt.Errorf("an error occured during unmarshal of authorization body: %v", err)
+	}
+
+	return nil
+}
+
+// createSubscription uses the official 'ctrlX Data Layer API' to create the sse subscription
+func (s *CtrlXDataLayerSSE) createSubscription(sub *subscription) (string, error) {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+
+	sseURL := combineScheme(s.Hostname) + "/automation/api/v2/events"
+
+	pl := sub.createRequestBody()
+	p, err := json.Marshal(pl)
+	if err != nil {
+		return "", fmt.Errorf("failed to create subscription %v payload, error: %v", sub.index, err)
+	}
+
+	requestBody := bytes.NewBuffer(p)
+	req, err := http.NewRequest("POST", sseURL, requestBody)
+	if err != nil {
+		return "", fmt.Errorf("failed to create post request to create sse subscription %v, error: %v", sub.index, err)
+	}
+
+	req.Header.Add("Authorization", s.token.String())
+
+	resp, err := s.connection.Do(req)
+	if err != nil {
+		return "", fmt.Errorf("failed to do post request to create sse subscription %v, error: %v", sub.index, err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 && resp.StatusCode != 201 {
+		return "", fmt.Errorf("failed to create sse subscription %v, status: %v", sub.index, resp.Status)
+	}
+
+	return sseURL + "/" + pl.Properties.ID, nil
+}
+
+// reconnect is called when the connection is broken
+// reconnect starts restablishing the connection
+func (s *CtrlXDataLayerSSE) reconnect(sub subscription) {
```

Comment:
Hmm the more I look at it, the more I think you should do the complete `reconnect` / `request` thing in _one_ go routine so you don't run into synchronization issues... There is no inter-dependency between subscription, is there?